### PR TITLE
Remove generation of instance trait by decl_storage.

### PR DESCRIPTION
**Pull request by thiolliere**
Because it is confusing and no longer necessary in any way, we can remove the declaration of instance trait in decl_storage.

DONE:
* decl_storage still generate Instance implementation for structs: Instance0, Instance1, ..., DefaultInstance and __InherentHiddenInstance.
* frame-support::traits::Instance is still reexported from inside the module: `pub use frame_support::traits::Instance as Instance;` to avoid any breaking change
* the Instance trait used in macro must be `Instance` (before it could be named whatever user wanted), this is to ensure people don't get confused about where it comes from. (if user use InstantiableThing for trait name, then an error message is returned).

Instance stuff are still not perfect but I hope it improves on the current situation.

To summarize the changes here:
before:
```rust
mod moduleA {
    decl_storage!(... for Module<T: Trait<I>, I: Instance) //<-- generate Trait Instance
}
mod moduleB {
    decl_storage!(... for Module<T: Trait<I>, I: Instance) //<-- generate another Trait Instance
}
```
after:
```rust
mod frame_support::traits { pub trait Instance { PREFIX .... }) //<-- declare the trait instance for all modules
mod moduleA {
    decl_storage!(... for Module<T: Trait<I>, I: Instance) //<-- do generate `pub use frame_support::traits::Instance;` to avoid breaking change)
}
mod moduleB {
     decl_storage!(... for Module<T: Trait<I>, I: Instance) //<-- do generate `pub use frame_support::traits::Instance;` to avoid breaking change)
}
```

### CODE REVIEW

File: frame/support/procedural/src/storage/instance_trait.rs [hunk 1]
```diff
@@ -70,42 +69,40 @@ pub fn decl_and_impl(scrate: &TokenStream, def: &DeclStorageDefExt) -> TokenStre
...
-
-	quote! {
-		/// Tag a type as an instance of a module.
-		///
-		/// Defines storage prefixes, they must be unique.
-		#optional_hide
-		pub trait #instance_trait: 'static {
-			/// The prefix used by any storage entry of an instance.
-			const PREFIX: &'static str;
-		}
```
Comment (thiolliere) [inline frame/support/procedural/src/storage/instance_trait.rs:108]: note: __GeneratedInstantiable is no longer generated because it was supposed to be only internally used (and so all our internal use now use frame_support::traits::Instance directly)

File: frame/support/procedural/src/storage/genesis_config/mod.rs [hunk 2]
```diff
@@ -104,10 +104,9 @@ fn impl_build_storage(
 		let name = syn::Ident::new(DEFAULT_INSTANCE_NAME, Span::call_site());
 		quote!( #name )
 	});
-	let inherent_instance_bound = def.optional_instance_bound.clone().unwrap_or_else(|| {
-		let bound = syn::Ident::new(DEFAULT_INSTANTIABLE_TRAIT_NAME, Span::call_site());
-		quote!( #inherent_instance: #bound )
-	});
+	let inherent_instance_bound = quote!(
+		#inherent_instance: #scrate::traits::Instance
+	);
```
Comment (thiolliere) [inline frame/support/procedural/src/storage/genesis_config/mod.rs:109]: here we use frame_support::traits::Instance directly instead of the internally created instance trait.

File: frame/support/procedural/src/storage/instance_trait.rs [hunk 3]
```diff
@@ -24,7 +24,6 @@ use super::DeclStorageDefExt;
 
 const NUMBER_OF_INSTANCE: usize = 16;
 pub(crate) const INHERENT_INSTANCE_NAME: &str = "__InherentHiddenInstance";
-pub(crate) const DEFAULT_INSTANTIABLE_TRAIT_NAME: &str = "__GeneratedInstantiable";
```
Comment (thiolliere) [inline frame/support/procedural/src/storage/instance_trait.rs:27]: note: __GeneratedInstantiable is no longer generated because it was supposed to be only internally used (and so all our internal use now use frame_support::traits::Instance directly)

File: frame/support/procedural/src/storage/instance_trait.rs [hunk 4]
```diff
@@ -70,42 +69,40 @@ pub fn decl_and_impl(scrate: &TokenStream, def: &DeclStorageDefExt) -> TokenStre
...
-		.unwrap_or_else(|| syn::Ident::new(DEFAULT_INSTANTIABLE_TRAIT_NAME, Span::call_site()));
-
-	let optional_hide = if def.module_instance.is_some() {
-		quote!()
+	if let Some(i) = def.module_instance.as_ref() {
+		let instance_trait = &i.instance_trait;
+		quote!(
+			/// Reexport from support::traits.
+			// This reexport is not strictly needed but used in order not to have breaking change.
+			pub use #scrate::traits::Instance as #instance_trait;
```
Comment (thiolliere) [inline frame/support/procedural/src/storage/instance_trait.rs:102]: here we reexport to avoid breaking change for user (before the trait Instance was generated here so it was in scope)

Like when user did `pub trait Trait<I: Instance = DefaultInstance {..}` in trait definition, the Instance trait was in scope.
If we don't reexport here, then user code will break and user have to write: `use frame_support::traits::Instance` themself.

File: frame/support/src/traits.rs [hunk 5]
```diff
@@ -1653,6 +1653,16 @@ impl<T> IsType<T> for T {
 	fn into_mut(&mut self) -> &mut T { self }
 }
 
+/// A type that holds an instance, it is used to defines the module prefix to be used by storages,
+/// thus it must unique for each module.
```
Comment (bkchr) [inline frame/support/src/traits.rs:1657]:
```suggestion
/// An instance of a pallet in the storage.
///
/// It is required that these instances are unique, to support multiple instances per pallet in the same runtime!
```

File: frame/support/procedural/src/storage/instance_trait.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thiolliere) [reply frame/support/procedural/src/storage/instance_trait.rs:102]: there is two reexport here: local reexport and public reexport.

Indeed if we don't mind breaking public reexport we can do `use #scrate::traits::Instance as Instance;` (e.g. nobody would have included the trait somewhere), I'm ok with this.

But for local reexport all instantiable crate currently use it.
E.g. when in pallet-balance we do `pub trait Trait<I: Instance = DefaultInstance>` we use local reexport.

Are you ok with only local use ?

```suggestion
			/// Local import of frame_support::traits::Instance
			// This import is not strictly needed but made in order not to have breaking change.
			use #scrate::traits::Instance as #instance_trait;
```

[the remark this answers is not in the document]
Comment (bkchr) [reply frame/support/procedural/src/storage/instance_trait.rs:102]: What do you mean by local re-export?

[the remark this answers is not in the document]
Comment (thiolliere) [reply frame/support/procedural/src/storage/instance_trait.rs:102]: just a local import: `use #scrate::traits::Instance`

[the remark this answers is not in the document]
Comment (thiolliere) [reply frame/support/procedural/src/storage/instance_trait.rs:102]: what I meant is that in your suggestion you do `use #scrate::traits::Instance as _;`
But this would break module doing just `pub trait Trait<I: Instance = DefaultInstance>`
Instead, if we want not to break, we should do `use #scrate::traits::Instance`

[the remark this answers is not in the document]
Comment (bkchr) [reply frame/support/procedural/src/storage/instance_trait.rs:102]: Ahh, I'm an idiot xD Yeah that is right :) Sorry. Sounds good to me.

**Comment by thiolliere** [comment]
bot merge
